### PR TITLE
only migrate events that were added via the proxy

### PR DESCRIPTION
**Pull request by BelfordZ**
Added new test cases to validate the enhanced functionality of the EventEmitter proxy. These tests check behaviors like event migration only for events added via the proxy, exclusion of internally added events, and proper handling of 'off' and 'removeListener' methods.

Changed Significant updates to the logic of creating an EventEmitter proxy. The changes focus on ensuring that only events added through the proxy are migrated when switching targets. This involves tracking events added to the proxy and only migrating events that were tracked in this manner, thus avoiding migrating events that were not added via the proxy.

In summary, The logic of this PR aims to refine the EventEmitter proxy's behavior, particularly in how events are handled and migrated, ensuring more precise and controlled event management.

This also fixes an issue with/unblocks https://github.com/MetaMask/metamask-extension/pull/22202

### CODE REVIEW

File: src/createEventEmitterProxy.ts [hunk 1]
```diff
@@ -93,8 +107,40 @@ export function createEventEmitterProxy<Type extends EventEmitterLike>(
...
+            name === 'once'
+          ) {
+            if (eventFilter(args[0])) {
+              eventsAdded.push({
+                addedWith: name,
+                name: args[0],
+                handler: args[1],
+              });
+            } else {
+              return target;
```
Comment (mcmire) [inline src/createEventEmitterProxy.ts:134]: Hmm, do we want to return the `target` here or return the `value`? It seems like we need tests to handle that an event that's being filtered out can still be listened to via `on`, `addListener`, `prependListener`, and `once`.

``` suggestion
              // This function may be either bound to something or nothing.
              // eslint-disable-next-line no-invalid-this
              return value.apply(this === receiver ? target : this, args);

File: src/createEventEmitterProxy.ts [hunk 2]
```diff
@@ -93,8 +107,40 @@ export function createEventEmitterProxy<Type extends EventEmitterLike>(
...
+            }
+          }
+          if (name === 'off' || name === 'removeListener') {
+            eventsAdded = eventsAdded.filter(
+              (addedEvent) =>
+                args[0] !== addedEvent.name || args[1] !== addedEvent.handler,
+            );
```
Comment (mcmire) [inline src/createEventEmitterProxy.ts:141]: Can we say:

``` suggestion
            removeEvent(args[0], args[1]);

File: src/createEventEmitterProxy.ts [hunk 3]
```diff
@@ -10,6 +10,20 @@ type EventEmitterLike = {
...
+  once(name: string, handler: (...args: unknown[]) => unknown): void;
+};
+
+type EventDetails = {
+  // the name of the event
+  name: string;
+  // the method given to handle the event
+  handler: (...args: unknown[]) => unknown;
+  // name of the method used to add the event (on or prepend)
+  addedWith: 'on' | 'prependListener' | 'addListener' | 'once';
```
Comment (mcmire) [inline src/createEventEmitterProxy.ts:26]: Nit: Can we use JSDoc for these so they'll show up on hover?

``` suggestion
  /**
   * The name of the event.
   */
  name: string;
  /**
   * The method given to handle the event.
   */
  handler: (...args: unknown[]) => unknown;
  /**
   * Name of the method used to add the event.
   */
  addedWith: 'on' | 'prependListener' | 'addListener' | 'once';

File: src/createEventEmitterProxy.ts
Comment (mcmire) [inline src/createEventEmitterProxy.ts:1]: There's some JSDoc above `createEventEmitterProxy` — is there anything we need to change about the description given these changes? Additionally, we might want to tweak the doc for the `Type` type parameter given that we've updated EventEmitterLike (could simplify it if you like since there are so many things supported now).

**Review by mcmire — COMMENTED** [review]
Great job tracking this down. The only thing I'm really wondering about is what we should do if an event being listened to is filtered out, but everything else looks good to me.

File: src/createEventEmitterProxy.ts (hunk 1 above)
Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:134]: Perhaps easiest to drop this block, and use `else if` for the `off`/`removeListener` check. Then it always falls through to the final `value.apply(this === receiver ? target : this, args);`

File: src/createEventEmitterProxy.test.ts [hunk 4]
```diff
@@ -31,9 +46,13 @@ describe('createEventEmitterProxy', () => {
...
 
+    original.emit('event');
     proxy.setTarget(next);
     next.emit('event');
     next.emit('event');
+    proxy.setTarget(original);
+    next.emit('event');
+    original.emit('event');
 
     expect(sawEvent).toBe(1);
```
Comment (jiexi) [inline src/createEventEmitterProxy.test.ts:57]: in this case it's also important when the event is seen. What are your thoughts on adding another one of these expects after the `original.emit('event');` on line 49?

File: src/createEventEmitterProxy.test.ts [hunk 5]
```diff
@@ -160,4 +179,63 @@ describe('createEventEmitterProxy', () => {
...
+  it('only migrates events that were added via the proxy, not all the events', () => {
+    const original = new EventEmitter();
+    const next = new EventEmitter();
+
+    let moveCount = 0;
+    original.on('shouldNotMove', () => (moveCount += 1));
+
+    const proxy = createEventEmitterProxy(original);
+    expect(proxy.eventNames()[0]).toBe('shouldNotMove');
+    proxy.on('shouldMove', () => (moveCount += 1));
```
Comment (jiexi) [inline src/createEventEmitterProxy.test.ts:192]: thoughts on using a new variable to keep track instead of reusing `moveCount`?

File: src/createEventEmitterProxy.test.ts [hunk 6]
```diff
@@ -160,4 +179,63 @@ describe('createEventEmitterProxy', () => {
...
+    proxy.setTarget(next);
+    next.emit('shouldMove');
+    expect(moveCount).toBe(1);
+    next.emit('shouldNotMove');
+    expect(moveCount).toBe(1);
+  });
+
+  it('does not migrate events that have been removed via "off"', () => {
```
Comment (jiexi) [inline src/createEventEmitterProxy.test.ts:200]: should test `removeListener` as well?

File: src/createEventEmitterProxy.test.ts (hunk 5 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.test.ts:192]: should make things a bit more clear

File: src/createEventEmitterProxy.test.ts (hunk 6 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.test.ts:200]: `off` is an alias to `removeListener` so I didn't test it, but since we are handling removeListener & off specifically, adding something to make sure theres no regression would be wise.

File: src/createEventEmitterProxy.ts
Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:1]: good call. I will revisit the JSDoc for createEventEmitterProxy after I address the other feedback.

File: src/createEventEmitterProxy.ts (hunk 3 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:26]: hmm it should be the same thing no?

File: src/createEventEmitterProxy.ts (hunk 1 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:134]: Hmm maybe I've misunderstood the purpose of the eventFilter. I was under the impression that when we filter an event, we are not wanting to events of that type. Given that, why would we want to add the subscription if the event is supposed to be filtered out?

The reason I return target here is because the return value when calling `on` (or equivalent) is always the event emitter itself (which is `target` here).

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:140]: `eventFilter` shouldn't suppress any events. The filtered events are excluded from _migration_, i.e. they don't get migrated when the target changes.

The code looks correct in that respect though. You're using the event filter correctly here I think.

>The reason I return target here is because the return value when calling on (or equivalent) is always the event emitter itself (which is target here).

This is the getter though, so you'd be returning the event emitter itself for `eventEmitter.on`. We'd want the `on` function to be returned instead, no? Maybe I'm missing something.

Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:140]: Ah thank you for clarifying. In this case, I believe my implementation is incorrect. Specifically, consider this code:
```
       if (
            name === 'on' ||
            name === 'addListener' ||
            name === 'prependListener' ||
            name === 'once'
          ) {
            if (eventFilter(args[0])) {
              eventsAdded.push({
                addedWith: name,
                name: args[0],
                handler: args[1],
              });
            } else {
              return target;
            }
          }
          if (name === 'off' || name === 'removeListener') {
            removeEvent(args[0], args[1]);
          }

          // This function may be either bound to something or nothing.
          // eslint-disable-next-line no-invalid-this
          return value.apply(this === receiver ? target : this, args);
        };
```

Assume the following:
 1. we call `proxy.on('foo', () => true)`. 
 2. we have eventFilter list ['foo']
 
 With our code, `value.apply(...)` would not be called, and thus the subscription would never be made.
 
 According to what you've explained here, this is incorrect. It should still create the subscription, it just should not migrate the subscription when the target is changed.

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:140]: Right, exactly. And this would be fixed by the suggestions Elliot and I left in the first two comments, I think.

**Comment by adonesky1** [comment]
@BelfordZ could you add a brief PR description?

**Comment by BelfordZ** [comment]
@adonesky1 done! I had written it just forgot to his save

File: src/createEventEmitterProxy.ts [hunk 7]
```diff
@@ -93,8 +121,33 @@ export function createEventEmitterProxy<Type extends EventEmitterLike>(
...
+          if (
+            name === 'on' ||
+            name === 'addListener' ||
+            name === 'prependListener' ||
+            name === 'once'
+          ) {
+            eventsAdded.push({
+              addedWith: name,
+              name: args[0],
+              handler: args[1],
```
Comment (Gudahtt) [inline src/createEventEmitterProxy.ts:144]: For `once`, should we be storing the unwrapped handler here? If someone tries to remove that listener, they'd be passing in the unwrapped handler, not the wrapped one. So the comparison in `removeEvent` would always fail to remove it.

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:144]: It would be great to add a test case for this as well (i.e. add an event listener using `once`, try to remove it, then confirm that it has been removed and hasn't been migrated)

File: src/createEventEmitterProxy.test.ts [hunk 8]
```diff
@@ -160,4 +203,93 @@ describe('createEventEmitterProxy', () => {
...
+    const proxy = createEventEmitterProxy(original);
+    const mockSetTarget = jest.fn();
+    proxy.setTarget = mockSetTarget;
+
+    const next = new EventEmitter();
+    proxy.setTarget(next);
+    expect(mockSetTarget).toHaveBeenCalledWith(next);
+  });
+
+  it('can handle the context being the proxy itself', () => {
```
Comment (Gudahtt) [inline src/createEventEmitterProxy.test.ts:285]: Could you clarify what the goal of this test is?

It looks like we're calling `on` and explicitly setting the `thisArg` to the context for the test itself. But the expectation is just checking what `this` is within the listener, which seems unrelated to the `thisArg` used for the call

Comment (Gudahtt) [reply src/createEventEmitterProxy.test.ts:285]: It definitely seems worth checking that `this` is set correctly for unbound listeners, but perhaps we can make this test more straightforward by removing step where `call` is used to change `thisArg`. But perhaps I'm missing something.

Comment (BelfordZ) [reply src/createEventEmitterProxy.test.ts:285]: This is just to get coverage on the line `return value.apply(this === receiver ? target : this, args);`
Im not sure why we needed to handle this, but someone added it and didnt test it, and it seemed easy enough to at least add a test that hits that line. Feel free to make a new issue about adding a better description / better test case for this, and maybe someone with more context around this ternary can write a better test for it.

File: src/createEventEmitterProxy.ts (hunk 7 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:144]: ah yes nice catch!

Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:150]: Okay I've updated PR accordingly. Lmk what you think

Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:150]: (this test should be what you were highlighting https://github.com/MetaMask/swappable-obj-proxy/pull/53/commits/578a5a0a7ffc3a8ad3f7da23c89564511e67d8ab#diff-5959fa448c7362c6ed2d30b45343e71a7a4e13358ea4fde75e9871e0b3fdad2eR271)

File: src/createEventEmitterProxy.test.ts (hunk 8 above)
Comment (Gudahtt) [reply src/createEventEmitterProxy.test.ts:311]: Ah darn, sorry I was getting mixed up, I was thinking of how the event listeners are always bound to the event emitter. But this isn't the listener, this is the call to add one. This all makes sense, nevermind.

Comment (BelfordZ) [reply src/createEventEmitterProxy.test.ts:311]: ahh gotcha. Yeah... NGL its pretty confusing up in here. Hard to keep track of what is what. Might be good to heavily JSdoc this puppy in a followup.

File: src/createEventEmitterProxy.ts (hunk 7 above)
Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:152]: Nice! Looks like it should work now, and the test coverage looks good

File: src/createEventEmitterProxy.ts [hunk 9]
```diff
@@ -93,8 +128,43 @@ export function createEventEmitterProxy<Type extends EventEmitterLike>(
...
+              filtered: !eventFilter(args[0]),
+            });
+          } else if (name === 'off' || name === 'removeListener') {
+            const eventAdded = eventsAdded.find(
+              ({ name: addedName, unwrappedHandler: addedUnwrappedHandler }) =>
+                addedName === args[0] && addedUnwrappedHandler === args[1],
+            );
+            // this should never really happen unless you've called removeListener for something that is already not there.
+            if (eventAdded === undefined) {
+              return target;
```
Comment (Gudahtt) [inline src/createEventEmitterProxy.ts:162]: HJmm, this doesn't seem quite right. If `removeListener` is called with an event that we don't know about, I think we should pass it on down through to the `value.apply` call below.

Returning `target` here is a bit strange because `target` is the underlying event emitter, but this function is the getter. So it means that something like `proxy.off('foo', handler)` would get the event emitter for `off`, making it `eventEmitter('foo', handler)`, but `eventEmitter` isn't a function, so that's not right.

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:162]: Oh I'm sorry, we're within the returned function, not within the `getter`. My bad, this is not doing what I thought.

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:162]: I see, and we'd want to avoid passing it down just in case that exact same event and listener was added via another proxy. That makes sense.

**Review by Gudahtt — APPROVED** [review]
LGTM, great work!

File: src/createEventEmitterProxy.ts (hunk 9 above)
Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:162]: Hmm yeah I couldnt decide on if it made more sense to just send the removeListner down or not. But I think you are probably right, sending it down makes more sense. In the event you are removing a listener that was not added via the proxy, but is present on the event emitter, the question was 'should we still remove it?' I talked myself in-and-out of it a few times over... But yeah I think you are right, we should send it.

As for the return value, This return is not the getter, it is essentially a wrapped version of the function returned by the getter. As such, the expected return value after calling `eventEmitter.removeListener('abc', () => 123)` is the eventEmitter itself. To be more specific to your example, the getter which returns a function for  `proxy.off` is still intact - we return a function. Its the return value of this function that we are changing.

Comment (BelfordZ) [reply src/createEventEmitterProxy.ts]: oaky updated, take a look!

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts]: Sorry for the additional mixup here! See my follow-up two comments, I think the way you have it here is actually safer in the case where two proxies point to the same target.

e.g. Lets say that `a` proxies `target`, and `b` proxies `target`.

If I run this code:

```
const handleNetworkChanged = () => console.log('network changed');
a.on('networkChanged', handleNetworkChanged)

b.on('networkChanged', handleNetworkChanged)

a.off('networkChanged', handleNetworkChanged)
a.off('networkChanged', handleNetworkChanged) // duplicate call to `off` here

target.emit('networkChanged')
```

Then we'd want `handleNetworkChanged` to still be called because `b` registered it and never unregistered it.

So your code is correct here, but if we passed it through like I suggested originally, this case would be broken

Comment (Gudahtt) [reply src/createEventEmitterProxy.ts]: Maybe that case would make for a good test as well, that's a pretty subtle scenario. Doesn't seem that unlikely either. I can see that happening if the thing we want to trigger is not directly related to either event emitter.

e.g. what if we want to trigger saving state if a network status changes for an "active" network, that might show up as this kinda scenario, where different proxies trigger the same persist call.

Comment (BelfordZ) [reply src/createEventEmitterProxy.ts:162]: Ive reverted the last two commits. If its okay with you, maybe we can create an issue to create test cases for 'many-to-one proxies to eventEmitter' - I suspect there might be a few nuanced items to figure out there.

**Review by Gudahtt — APPROVED** [review]

File: src/createEventEmitterProxy.ts (hunk 9 above)
Comment (Gudahtt) [reply src/createEventEmitterProxy.ts:162]: Sure, sounds good
